src/editor/editor.cpp

```diff
@@ -0,0 +1,361 @@
+  }
+
+  // update other stuff
+  if (levelloaded) {
+    currentsector->update(0);
+    tileselect.update(elapsed_time);
+    layerselect.update(elapsed_time);
+    inputcenter.update(elapsed_time);
+    update_keyboard();
+  }
+}
+
+void Editor::test_level() {
+  leveltested = true;
+  Tile::draw_editor_images = false;
+  level->save("levels/misc/test.stl");
+  std::unique_ptr<World> test_world = World::load("levels/misc");
+  GameManager::current()->start_level(std::move(test_world), "test.stl");
+}
+
+bool Editor::can_scroll_vert() {
+  return currentsector->get_height() + 32 > SCREEN_HEIGHT;
+}
+
+bool Editor::can_scroll_horz() {
+  return currentsector->get_width() + 128 > SCREEN_WIDTH;
+}
+
+void Editor::scroll_left(float speed) {
+  if (can_scroll_horz()) {
+    if (currentsector->camera->get_translation().x >= speed*32) {
+      currentsector->camera->move(-32 * speed, 0);
+    } else {
+      //When is the camera less than one tile after the left limit, it puts the camera to the limit.
+      currentsector->camera->move(-currentsector->camera->get_translation().x, 0);
+    }
+    inputcenter.actualize_pos();
+  }
+}
+
+void Editor::scroll_right(float speed) {
+  if (can_scroll_horz()) {
+    if (currentsector->camera->get_translation().x <= currentsector->get_width() - SCREEN_WIDTH + 128 - 32 * speed) {
+      currentsector->camera->move(32*speed, 0);
+    } else {
+      //When is the camera less than one tile after the right limit, it puts the camera to the limit.
+      // The limit is shifted 128 pixels to the right due to the input gui.
+      currentsector->camera->move(
+            currentsector->get_width() - currentsector->camera->get_translation().x - SCREEN_WIDTH +128, 0);
+    }
+    inputcenter.actualize_pos();
+  }
+}
+
+void Editor::scroll_up(float speed) {
+  if (can_scroll_vert()) {
+    if (currentsector->camera->get_translation().y >= speed*32) {
+      currentsector->camera->move(0,-32*speed);
+    } else {
+      //When is the camera less than one tile after the top limit, it puts the camera to the limit.
+      currentsector->camera->move(0, -currentsector->camera->get_translation().y);
+    }
+    inputcenter.actualize_pos();
+  }
+}
+
+void Editor::scroll_down(float speed) {
+  if (can_scroll_vert()) {
+    if (currentsector->camera->get_translation().y <= currentsector->get_height() - SCREEN_HEIGHT - 32 * speed) {
+      currentsector->camera->move(0, 32*speed);
+    } else {
+      //When is the camera less than one tile after the bottom limit, it puts the camera to the limit.
+      // The limit is shifted 32 pixels to the bottom due to the layer toolbar.
+      currentsector->camera->move(0,
+            currentsector->get_height() - currentsector->camera->get_translation().y - SCREEN_HEIGHT +32);
+    }
+    inputcenter.actualize_pos();
+  }
+}
+
+void Editor::update_keyboard() {
+
+  if (!enabled){
+    return;
+  }
+
+  if (InputManager::current()->get_controller()->pressed(Controller::ESCAPE)) {
```

Comment:
Can you do something like
auto controller = InputManager::current()->get_controller();

and then replace the relevant occurrences in the following lines?